### ISSUE TITLE
Proposal: Pattern Matching

### ISSUE THREAD

**Issue by b-studios** [effekt-lang/effekt#383]
_this is still work in progress_

Here I propose quite a lot of changes to pattern matching and control flow structures (`if` and `while`). All of the changes will happen in the front end. In particular, the representation of core and all backends should stay largely unaffected.

## Current Implementation
We currently implement pattern matching in the style of [Jules Jacobs](https://julesjacobs.com/notes/patternmatching/patternmatching.pdf).

From an implementors perspective, this is particularly attractive since compiling pattern matching automatically gives us exhaustiveness and redundancy checks. Pattern matching is compiled in the translation form [source to core](https://github.com/effekt-lang/effekt/blob/3634fe2cda7962f7e3910a7400cdec282036df75/effekt/shared/src/main/scala/effekt/core/Transformer.scala#L612).

https://github.com/effekt-lang/effekt/blob/3634fe2cda7962f7e3910a7400cdec282036df75/effekt/shared/src/main/scala/effekt/core/Transformer.scala#L579-L590

The internal representation of a single clause is the _conjunction_ of potentially several scrutinees  with the expected pattern, followed by the clause to be executed in case the patterns are all satisfied.

## Current Limitations

User-facing, we currently only support matching on data type constructors, such as

```scala
e match {
  case Some(v) => ...
  case None => ...
}
```

In particular, we do not implement
- guards (that is, `case Some(v) if v > 0 => ...`)
- as-patterns (that is `case Some(v) as x => ...` or `case x @ Some(v) => ...`
- literal patterns (that is `case 1 => ...`).

As @phischu noted in #363, we can easily support literal patterns if we support guards.

Patterns, currently, can only occur in a match expression. Some languages (like Rust's [while let](https://doc.rust-lang.org/rust-by-example/flow_control/while_let.html)) support matching as part of conditionals such as:

```scala
while (find(x) is Some(v)) {
  ...
}
```

This is sometimes more efficient, since it prevents us from having to call `find` twice
```scala
while (find(x).isDefined) {
 ...  find(x).get ....
}
```
Of course we can always write this explicitly as a recursive function, which is more noisy:
```scala
def loop(): Unit = find(x) match {
  case Some(v) =>  ...; loop()
  case None() => ()
}
loop()
```

## Proposal
This proposal aims to address the following issues:

- Support literal patterns
- Support guards
- Support matching in conditionals (while and if)

while retaining most of the simple implementation style of Jules.

It takes inspiration from the ultimate conditional syntax (UCS) by [Lionel Parreaux](https://cse.hkust.edu.hk/~parreaux/ucs_v1.2.pdf). However, importantly, I do not propose to fully adopt Lionel's proposal but moderate it slightly.

### Concepts and Changes
We introduce the concept of a **guard** that can be either a boolean expression or a (partial) match:

```
guard ::= e is pattern     (partial match)
        | e                (of type Boolean)         
```

A **pattern** is almost defined as before, only that we now also allow binding the scrutinee:

```
pattern ::= C([pattern]*)     (constructor pattern)
          | x                 (name pattern)
          | _                 (wildcard pattern)
```

As before a **`match` expression** always destructs the receiver (in the future also multiple receivers, comma separated) into cases to be considered. The cases are dealt with in separate **clauses**:

```
clause ::= case pattern [and guard]* => s
```
Each clause now also can contain multiple guards, where each guard (and the statement `s`) can refer to the variables bound by previous patterns.

Statements still contain `match`, `if` and `while`, but all three 

1. now support guards
2. include `else` branches

```
s ::= e match { [clause]* } [else { s }]?
    | if (guard [and guard]*) { s } [else { s }]?
    | while (guard [and guard]*) { s } [else { s }]?
```

Here, the `else` branch can be seen as a "handler" for the partial match.

Finally, we will desugar literal patterns into guards.
```scala
e match {
  case 1 => ...
  case _ => ...
}
// becomes
e match {
  case x and x == 1 => ...
  case _ => ...
}
```

### Consequence 1
We now allow partial patterns as part of conditionals

```scala
if (find(x) is Some(v)) {
  ... v ...
} else {
  ...
}
```
Same also holds for `while` which additionally now supports `else` clauses:

```scala
while (find(x) is Some(v)) {
  ... v ...
} else {
  ...
}
```

That is, the predicate of a conditional (now called a "guard") can either be a boolean value or a partial match. This should not be surprising, since conceptually, treating booleans as a datatype with constructors `True` and `False`
```scala
if (x > 0) { ... } else { ... }
```
corresponds to
```scala
if (x > 0 is True) { ... } else { ... }
```
even though Booleans are special cased.

### Consequence 2
The support of guards with `and` is illustrated in this an example from `seq` again:

```scala
def unconsFinger[A](seq: Tree[A]): TreeView[A] = seq match {
  case NoFinger()      => TreeIsEmpty()
  case SingleFinger(v) => TreeView(v, NoFinger())
  case DeepFinger(size, One(a), m, s) and unconsFinger(m) is TreeView(b, m2) => 
    TreeView(a, DeepFinger(size - a.size, b.toDigitNode, m2, s))
  case DeepFinger(size, One(a), m, s) => TreeView(a, s.toTree)
  // ... more cases
}
```

Here, the left-hand-side of `is` is always an arbitrary expression (maybe restrict it to be pure?) that can refer to variables bound by the preceding pattern.

### Consequence 3
We now support `as`-patterns to name intermediate scrutinees.
```scala
e match {
  case Cons(Some(v) as x, rest) => ...
}
```

## Implementation
- Our implementation already features "and" patterns, but we do not make use of it. 
- Desugaring guards should happen before or in Namer since this way we can already look for the correct implementation of `infixEq`. Doing this in Typer or in transformation to core might be too late
- Allowing partial matches in conditional control flow structures might require some more additional changes that I cannot foresee right now.
- I don't know how adding guards will affect the exhaustiveness check that we get for free at the moment.

**Comment by b-studios** [comment]
@phischu @jiribenes @marzipankaiser @marvinborner @dvdvgt I would like to discuss this in our next Effekt working group meeting. It would be great if you could look into it (and maybe even briefly into https://julesjacobs.com/notes/patternmatching/patternmatching.pdf) before the meeting so that we can discuss this constructively.

**Comment by b-studios** [comment]
(updated the proposal, changed some design decisions, and added grammars to facilitate more concrete discussion)

**Comment by jiribenes** [comment]
Looks great, I'm looking forward to discussing this in the meeting :)

However, I'd really like to hear what's the plan for other possible features like ranges and or-patterns (as implemented in [Yorick Peterse's version](https://gitlab.com/yorickpeterse/pattern-matching-in-rust/-/tree/main/jacobs2021)). I'm not sure if such features are out of scope for this proposal or if they are not planned at all in the near future. 🤔

**Comment by b-studios** [comment]
If I understand correctly, he is implementing or-patterns by desugaring them early on. Do you have a concrete proposal for or-patterns and intervals? I would assume the latter could be desugared into guards as well, no?

**Comment by b-studios** [comment]
Another extension, not mentioned yet, are (user defined) splices. We already implement them for externs. We could potentially desugar

```scala
e match {
  case url"/user/${group}/${id}" => ...
}
```
to
```scala
e match {
  case x and url(List("/user/", "/")) is Cons(group, Cons(id, Nil())) => ...
}
```
or similar.

**Comment by BinderDavid** [comment]
Just as some food for thought, the following two pages from the GHC wiki discuss the design space for patterns quite a bit:
https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns
https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns-alternative
In discussions I have frequently seen the second page mentioned as a sweetspot for an extension to pattern matching which can express a lot of other pattern matching features. (Syntax is haskelly, but could easily be modified.)

**Comment by b-studios** [comment]
Thanks for chiming in @BinderDavid. If I understand correctly, the following use of a view pattern (mentioned in view-pattern-alternative)

```haskell
  length :: JList a -> Integer
  length (l | Nil <- view l) = 0
  length (l | Cons x xs <- view l) = 1 + length xs
```
could be expressed as
```scala
def length(l: JList[A]) = l match {
  case l and view(l) is Nil => 0
  case l and view(l) is Cons(x, xs) => 1 + length(xs)
}
```
right?

**Comment by BinderDavid** [comment]
Currently you support guards only in clauses, but not in the syntax of patterns itself. I.e. they only appear in this position in the grammar:
```
clause ::= case pattern [and guard]* => s
```
This means that guards cannot be used nested inside guards. In https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns-alternative this would be allowed, since the grammar would look somewhat like this:
```
pattern ::= C([pattern]*) [as x]?    (constructor pattern)
          | x                        (name pattern)
          | _                        (wildcard pattern)
          | pattern [and guard]+ (guarded pattern)
          
guard ::= e is pattern     (partial match) 
        | e                (of type Boolean)      
```
Was the omission of guards inside the `e is pattern` intended? I guess there could be parsing ambiguities which have to be resolved.

**Comment by b-studios** [comment]
I have three concerns with your proposal, @BinderDavid:

1. The grammar is ambiguous since the guarded pattern can be left-nested. This could be improved by
    ```
    pattern ::= C([pattern]*) [as x]?    (constructor pattern)
              | x                        (name pattern)
              | _                        (wildcard pattern)
              | pattern and guard       (guarded pattern)
    ```
    However, the nesting feels wrong this way. The outer most pattern now is the last guard.
  
2. The implementation that @jiribenes pointed at (in Rust) only supports one guard per _clause_. My proposal already generalizes this, but I don't know how well the approach scales to _nested_ guards.

3. Maybe this is difficult to understand for users and results in overly complicated code? The example you gave me in private communication
    ```scala
    case (Left(x) and lookup(x) is True) | (Right(y) and wurble(y) is False) => ...
    ```
    seems not too bad, though.
    
    
That said, we probably need it if we really want to desugar literal patterns, interval patterns, etc. into guards since those patterns _can_ occur freely in pattern position.

**Comment by b-studios** [comment]
The implementation is coming along. There are a few major things left:

- exhaustivity checking
- or-patterns
- reconsidering nested guards

I am inclined to not support or-patterns, since I don't know how to do that concisely (both Namer and PatternMatchingCompiler would become much more complicated).

Any opinions on this? Does anybody have a simple idea on how to support them?

**Comment by b-studios** [comment]
We can of course add more features later, such as

- Or Patterns
- Branching guards, that is
```
   case Some(v) and foo(v) match
      case Some(w) => ...
   case sharedDefault => ...
```
and more
